### PR TITLE
refactor: reworked Component's Disable and Enable methods, Serialization overriding simplified

### PR DESCRIPTION
**Pull request by gerard-queralt**
# Points of interest

- A Component is considered Active if and only if it is active and its owner is active
- Refactored Component::Enable and Component::Disable
    - Methods in base class are no longer virtual, now they call Component::SignalEnable and Component::SignalDisable, respectively. These virtual methods can be overridden and used to manage what needs to happen in such cases. This way, the Component's `enabled` member remains untouched.
- Refactored serialization methods (Save/Load) to follow a similar pattern: public method does commont serialization procedure (name, active, etc.) and calls a private, pure virtual method that each component will implement to manage their serialization.
- `GameObject::active` member has been removed from serialization, since it's a computed value anyways.

### CODE REVIEW

**Review by PaablooCH — CHANGES_REQUESTED** [review]
Disable by the imgui works (i cannot test it with my scripts, but i assume that will be working because they use the same functions), but the issue that you linked is not solved yet
image

**Comment by gerard-queralt** [comment]
> the issue that you linked is not solved yet

The problem is that we cannot just blanket disable all components, because then we would need to re-enable them one by one. We may need to take a look at how the physics simulation can handle these things, like a way of checking if the rigidbody is enabled. Better to talk with @OriolSors about this.

**Comment by PaablooCH** [comment]
> We may need to take a look at how the physics simulation can handle these things, like a way of checking if the rigidbody is enabled. Better to talk with @OriolSors about this.

It could be a missunderstanding, but what i'm selecting in that photo is a light. The light is still rendered even though is disabled

**Comment by DADSILVER** [comment]
I see this pr fine, but I miscommunicated what I need. Because right now we need to disable all the components from a game object. Then we need to make the component update only happen if it and its parent are active. 

If we created a func that checks if the obj or a parent of this is deactivated, we may manage from the updated component.

**Comment by gerard-queralt** [comment]
> right now we need to disable all the components from a game object.

Again, can't do that. We can only consider the component disabled, which is what I tried to achieve here. Changing the value of the member in Component is a bad idea. What is the expected behaviour, then, when a GameObject is enabled again? Enable _all_ components? Leave them all disabled?

> Then we need to make the component update only happen if it and its parent are active.

This is already being done, check [the code in develop](https://github.com/Horizons-Games/Axolotl-Engine/blob/develop/Source/Modules/ModuleScene.cpp#L136).

**Comment by markmeliaa** [comment]
> We can only consider the component disabled, which is what I tried to achieve here. Changing the value of the member in Component is a bad idea. What is the expected behaviour, then, when a GameObject is enabled again? Enable _all_ components? Leave them all disabled?

Can't we just, in the Update() of ModuleScene, call the update of those components whose owner GameObject is enabled/active? We would not have to enable/disable any component ourselves, just not update those which are in a GameObject that is not enabled/active.

**Comment by gerard-queralt** [comment]
> Can't we just, in the Update() of ModuleScene, call the update of those components whose owner GameObject is enabled/active?

Again, this is already happening; a component is only updated if it's enabled. The issue happens with stuff that has additional logic. If a light is disabled, we need to update and re-render them. If an audio source is disabled, we need to stop the Wwise process attached to it. For rigid bodies, we need to remove them from the Bullet simulation... Right now, all of this is done by _pushing_ this information to whoever needs to know it. If we _pulled_ this information we would need to check each of these components each frame, which I think is far from ideal. I tried to keep this sort of signaling we have with this refactor.

**Comment by markmeliaa** [comment]
But the answer can't just be "we can't do it". It is expected that, if anyone disables a GameObject, its components stop working as well, so we have to find a way to communicate back to the component when its Update() is stopped.

File: Source/DataModels/Windows/EditorWindows/WindowInspector.cpp [hunk 1]
```diff
@@ -176,7 +180,7 @@ void WindowInspector::InspectSelectedGameObject()
 	if (lastSelectedGameObject)
 	{
 		bool enable = lastSelectedGameObject->IsEnabled();
-		ImGui::Checkbox("Enable", &enable);
+		bool enableStateChanged = ImGui::Checkbox("Enable", &enable);
```
[the remark this answers is not in the document]
Comment (gerard-queralt) [reply Source/DataModels/Windows/EditorWindows/WindowInspector.cpp:183]: Basically, yes. We can avoid calling `Enable`/`Disable` each frame, which we were doing before.

File: Source/DataModels/GameObject/GameObject.h [hunk 2]
```diff
@@ -90,19 +90,19 @@ class GameObject
 
 	Component* CreateComponentLight(LightType lightType, AreaType areaType);
 
-	bool IsEnabled() const; // If the check for the GameObject is enabled in the Inspector
+	// This method returns true if and only if the GameObject is currently enabled
+	bool IsEnabled() const;
 	void Enable();
 	void Disable();
 
+	// This method returns true if IsEnabled returns true for this GameObject or any of its "ancestors" in the hierarchy
```
[the remark this answers is not in the document]
Comment (markmeliaa) [reply Source/DataModels/GameObject/GameObject.h:98]: Yes, that's the expected behaviour, but in the comment what you are saying is that, if either A or B are enabled, then C will have `IsActive` set to `true`, which is wrong.

File: Source/DataModels/Components/Component.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	{
+		enabled = false;
+	}
+	SignalDisable();
+}
+
+void Component::Save(Json& meta)
+{
+	meta["type"] = GetNameByType(type).c_str();
+	meta["active"] = (bool) enabled;
```
[the remark this answers is not in the document]
Comment (markmeliaa) [reply Source/DataModels/Components/Component.cpp:48]: Oh I haven't thought about that, good point. We can leave it at that by now ig.

File: Source/DataModels/GameObject/GameObject.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (markmeliaa) [reply Source/DataModels/GameObject/GameObject.h]: Np! I had to do the base of all this enable/active ordeal months ago so I am well aware that it is quite confusing at times.

**Comment by markmeliaa** [comment]
> What I think is a bad idea is to change the value of `enabled` (or `active`, before) in the component. I think that would be a very bad option from a UI standpoint. 

Agreed.

**Comment by markmeliaa** [comment]
I'm encountering an issue that, in the scene `GameplayFullTrial_VS3.axolotl` if I disable either Bix or any drone, press play, and try to enable them again, the engine breaks. Seems to be a problem with the AnimationComponent, here's the call stack of the error:
image

It seems that, from the scripts, we are trying to change the state of a ComponentAnimation that is disabled, so maybe we'll have to check in the script if any components that we use are enabled?

PS: Those GameObjects that do not have a ComponentAnimation work and get enable/disable correctly, such as the plane.

**Comment by gerard-queralt** [comment]
Found the issue. Since the script is disabled, we never call `Start`, so in this case the `animation` member is null. But this is obviously an engine-side issue, nothing wrong with the script itself. Do you think it'd be okay to always `Init` and `Start` scripts, regardless of if they are enabled or not? Not sure of the implications...

**Comment by markmeliaa** [comment]
>Do you think it'd be okay to always Init and Start scripts, regardless of if they are enabled or not?

Hmmm... I find that a better solution (although maybe complex to perform) could be that the scripts always call to their `Init`s and `Start`s methods when they get activated for the first time.

Instead of always executing the Init and Start regardless of them being active or not, we will just call the Init and Start WHEN they get activated for the first time (whenever that may be). This will be useful for, if in the future we have prefabs and we instantiate them mid-game, their scripts will always get Init and Start-ed before anything else.

**Comment by gerard-queralt** [comment]
Checking Unity's documentation, seems like there it behaves like a mix of both approaches. `Awake`, the equivalent of `Init`, is called even if the script is disabled (but, unlike what I suggested, it checks that the GameObject is enabled). `Start` is only called when the script first becomes active, like you suggested. We could mimic this behaviour, if you agree.
Documentation:
- https://docs.unity3d.com/ScriptReference/MonoBehaviour.Awake.html
- https://docs.unity3d.com/ScriptReference/MonoBehaviour.Start.html

**Comment by gerard-queralt** [comment]
Went ahead and did it how it's done in Unity. Let me know if you'd rather do it the way you suggested, since it's very similar and it won't require much change. At least now we have it working.

**Comment by markmeliaa** [comment]
> Went ahead and did it how it's done in Unity. Let me know if you'd rather do it the way you suggested, since it's very similar and it won't require much change. At least now we have it working.

Sorry for the late reply. I find mimicking Unity a good solution, so now any component that we would like to save through the scripts should be in the `Init()` and any additional logic should go in the `Start()`, gotcha.

Let me check real quick that everything works correctly so I can approve this.

**Review by markmeliaa — DISMISSED** [review]

**Comment by gerard-queralt** [comment]
>so now any component that we would like to save through the scripts should be in the `Init()` and any additional logic should go in the `Start()`, gotcha.

That's not necessarily true, I imagine it would depend on each script. You just need to keep in mind that `Init` will be called as long as the owner is enabled.

I don't know exactly how the scripts work, so I could fumble my example but anyway. Let's say we have a UI script that displays the health of the entity it's attached to. In order to do so, it finds the health system script in its owner and reads the current health. But, if this health system is disabled for some reason, that value won't be properly initialized, so the UI will also be incorrect. If you initialize the current health in the Init, you can know for sure that the read will be correct. This is specially important when sharing or passing pointers between scripts.

As a rule of thumb, I think you can follow Unity recommendations: if script A depends on script B, use `Init` to set the necessary values in B. If values are self-contained (private members that won't be read by anyone else), I don't think using `Init` or `Start` makes any difference.

**Review by markmeliaa — DISMISSED** [review]

**Review by markmeliaa — APPROVED** [review]

**Review by Dedado3000 — APPROVED** [review]

**Review by markmeliaa — APPROVED** [review]

**Pull request merged**